compiler/codegen/CCData.cpp

```diff
@@ -0,0 +1,144 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "omrcomp.h"
+#include "OMR/Bytes.hpp"
+
+#include "codegen/CCData.hpp"
+
+#include "infra/Monitor.hpp"
+#include "infra/CriticalSection.hpp"
+
+using OMR::CCData;
+
+CCData::key_t CCData::key(const uint8_t * const data, const size_t sizeBytes)
+   {
+   return key_t(reinterpret_cast<const key_t::value_type *>(data), sizeBytes);
```

Comment:
Is there `reinterpret_cast` guaranteed to be safe?

Reply:
If `key_t::value_type` boils down to a `char`, then yes. If `key_t::value_type` is something else, then not necessarily. A `wstring` key type, for example, wouldn't work, since you'd be casting to a non-byte pointer.

That's a useful observation actually, it implies that whatever type is chosen to represent the key, it has to be constructible from a sequence of bytes. So rather than the `static_assert(sizeof(key_t::value_type) == 1, ...)` elsewhere, we need the much stronger `static_assert(std::is_same<key_t::value_type, char>::value, ...)` to avoid breaking aliasing rules.